### ISSUE TITLE
Confusing reformatting of multiline function arguments with functions in them

### ISSUE THREAD

**Issue by Augustin82** [prettier/prettier#585]
Sorry about the title, I found out it's not so easy for me to describe the problem ;)

I encountered the problem using [reselect](https://github.com/reactjs/reselect)'s createSelector, but this applies to many use cases, in particular when using React/Redux and HoCs.

Original code:
```js
const selectRedirectionQuery = () => createSelector(
  selectRoute(),
  (routingState) => routingState.getIn(['locationBeforeTransitions', 'query', 'redirect'])
);
```

Output (options: ` --trailing-comma=true --bracket-spacing=true --single-quote=true`):
```js
const selectRedirectionQuery = () =>
  createSelector(selectRoute(), routingState =>
    routingState.getIn(['locationBeforeTransitions', 'query', 'redirect']));
```

Expected one of these, depending on the closure's length:
```js
const selectRedirectionQuery = () =>
  createSelector(
    selectRoute(),
    routingState =>
      routingState.getIn(['locationBeforeTransitions', 'query', 'redirect'])
  );
```
```js
const selectRedirectionQuery = () =>
  createSelector(
    selectRoute(),
    routingState => routingState.getIn(['locationBeforeTransitions', 'query', 'redirect'])
  );
```

The current output seems to me to be really hard to parse correctly.

**Comment by vjeux** [comment]
In JavaScript, an extremely common pattern is to have the last argument of a call to be a callback. Because of this, people tend to put the arguments part of the function inline and indent the body. This reduces the number of indentation.

Your case fits this pattern well. `createSelector` is configured by some attributes that fit in one line (`selectRoute()`) and has a callback with the new value (`routingState`). In the next line and indented, you've got the actual action (`routingState.getIn...`).

I think that the prettier output here is what we want and it's a matter of getting used to the writing style.

**Comment by Augustin82** [comment]
That's assuming only one of the arguments is a function. My experience is that it's often enough not the case.

**Comment by vjeux** [comment]
Thanks for reporting this but we are unlikely to change the behavior here. I'm going to close this issue but please feel free to open issues for other things that don't feel right to you :)
